internal/controller/core/vtep_controller.go

```diff
@@ -0,0 +1,427 @@
+			if err := r.Update(ctx, obj); err != nil {
+				log.Error(err, "Failed to remove finalizer from resource")
+				return ctrl.Result{}, err
+			}
+		}
+		log.Info("Resource is being deleted, skipping reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers
+	if !controllerutil.ContainsFinalizer(obj, v1alpha1.FinalizerName) {
+		controllerutil.AddFinalizer(obj, v1alpha1.FinalizerName)
+		if err := r.Update(ctx, obj); err != nil {
+			log.Error(err, "Failed to add finalizer to resource")
+			return ctrl.Result{}, err
+		}
+		log.Info("Added finalizer to resource")
+		return ctrl.Result{}, nil
+	}
+
+	orig := obj.DeepCopy()
+	if conditions.InitializeConditions(obj, v1alpha1.ReadyCondition) {
+		log.Info("Initializing status conditions")
+		return ctrl.Result{}, r.Status().Update(ctx, obj)
+	}
+
+	// Always attempt to update the metadata/status after reconciliation
+	defer func() {
+		if !equality.Semantic.DeepEqual(orig.ObjectMeta, obj.ObjectMeta) {
+			if err := r.Patch(ctx, obj, client.MergeFrom(orig)); err != nil {
+				log.Error(err, "Failed to update resource metadata")
+				reterr = kerrors.NewAggregate([]error{reterr, err})
+			}
+			return
+		}
+
+		if !equality.Semantic.DeepEqual(orig.Status, obj.Status) {
+			if err := r.Status().Patch(ctx, obj, client.MergeFrom(orig)); err != nil {
+				log.Error(err, "Failed to update status")
+				reterr = kerrors.NewAggregate([]error{reterr, err})
+			}
+		}
+	}()
+
+	if err = r.reconcile(ctx, s); err != nil {
+		log.Error(err, "Failed to reconcile resource")
+		return ctrl.Result{}, err
+	}
+
+	// force a periodic requeue to enforce state is in sync
+	return ctrl.Result{RequeueAfter: Jitter(r.RequeueInterval)}, nil
+}
+
+// vtepScope holds k8s objects used during a reconciliation.
+type vtepScope struct {
+	Device         *v1alpha1.Device
+	VTEP           *v1alpha1.VTEP
+	Connection     *deviceutil.Connection
+	ProviderConfig *provider.ProviderConfig
+	Provider       provider.VTEPProvider
+}
+
+func (r *VTEPReconciler) reconcile(ctx context.Context, s *vtepScope) (reterr error) {
+	if s.VTEP.Labels == nil {
+		s.VTEP.Labels = make(map[string]string)
+	}
+	s.VTEP.Labels[v1alpha1.DeviceLabel] = s.Device.Name
+
+	if !controllerutil.HasControllerReference(s.VTEP) {
+		if err := controllerutil.SetOwnerReference(s.Device, s.VTEP, r.Scheme, controllerutil.WithBlockOwnerDeletion(true)); err != nil {
+			return err
+		}
+	}
+
+	if s.ProviderConfig != nil {
+		if err := r.validateProviderConfigRef(ctx, s); err != nil {
+			conditions.Set(s.VTEP, metav1.Condition{
+				Type:    v1alpha1.ConfiguredCondition,
+				Status:  metav1.ConditionFalse,
+				Reason:  v1alpha1.IncompatibleProviderConfigRef,
+				Message: fmt.Sprintf("ProviderConfigRef is not compatible with Device: %v", err),
+			})
+			return fmt.Errorf("provider config reference is not compatible with the device: %w", err)
```

Comment:
These should use the `reconcile.TerminalError` https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/reconcile#pkg-functions such that they don't trigger a exponential backoff retrying (as there is no point in retrying). Also applies to the errors returned below.

I would directly set the conditions in the `validateX` funcs and then return a `reconcile.TerminalError` from those funcs and then within this `reconcile` func just `return err` as is.